simulator/src/main/scala/scienceworld/tasks/goals/Goal.scala

```diff
@@ -331,9 +333,82 @@ class GoalSequence(val subgoals:Array[Goal], optionalUnorderedSubgoals:Array[Goa
     os.toString()
   }
 
+  def toJSONArray(goals: Array[Goal], requiredGoals: Boolean): String = {
+    val jsonGoals = new ArrayBuffer[String]
+    for (i <- 0 until goals.length) {
+      val goal = goals(i)
+      val goalDescription = goal.description
+      val goalClass = goal.getClass.toString.split("\\.").last.formatted("%40s")
+      val key = goal.key
+      val keysMustBeCompletedBefore = goal.keysMustBeCompletedBefore
+      val keysMustBeCompletedBeforeJSON = "[" +
+        keysMustBeCompletedBefore.map(s => "\"" + GoalSequence.sanitizeJSON(s) + "\"").mkString(", ") + "]"
+      val satisfiedWithObject = goal.satisfiedWithObject
+      val satisfiedWithObjectJSON =
+        if (satisfiedWithObject.isDefined) {
+          "\"" +  GoalSequence.sanitizeJSON(satisfiedWithObject.get.getDescription(MODE_CURSORY_DETAIL)) + "\""
+        } else {"null"}
+      val defocusOnSuccess = goal.defocusOnSuccess
+      val isOptional = goal.isOptional
+      val passed =
+        if (requiredGoals) {
+          i < this.curSubgoalIdx
+        } else {
+          this.optionalUnorderedSubgoalsCompleted(i)
+        }
+
+      val json = new StringBuilder
+      json.append("{")
```

Comment:
Building JSON as a string is a good tradeoff between including a new library and doing the least amount of work possible. We can always come back and use a JSON serialization library, but this is perfectly good for now.

Reply:
I agree! I'm feel this kind of logic is quite brittle, but I've done enough testing and reasoning through the code that I'm pretty confident it doesn't have any serialization bugs.